java.lang.NullPointerException: Cannot read the array length because "<local2>" is null

Faced this issue once while playing with nosqlbench:

```
[cluster1-worker-0] ERROR com.datastax.cql3.scylla_shaded.driver.core.ExceptionCatchingRunnable - Unexpected error while executing task
java.lang.NullPointerException: Cannot read the array length because "<local2>" is null
    at com.datastax.cql3.scylla_shaded.driver.core.HostConnectionPool.closeAsync(HostConnectionPool.java:841)
    at com.datastax.cql3.scylla_shaded.driver.core.SessionManager.removePool(SessionManager.java:440)
    at com.datastax.cql3.scylla_shaded.driver.core.SessionManager.onDown(SessionManager.java:528)
    at com.datastax.cql3.scylla_shaded.driver.core.Cluster$Manager.onDown(Cluster.java:2207)
    at com.datastax.cql3.scylla_shaded.driver.core.Cluster$Manager.access$1200(Cluster.java:1560)
    at com.datastax.cql3.scylla_shaded.driver.core.Cluster$Manager$5.runMayThrow(Cluster.java:2162)
    at com.datastax.cql3.scylla_shaded.driver.core.ExceptionCatchingRunnable.run(ExceptionCatchingRunnable.java:32)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at com.datastax.internal.com_google_common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:108)
    at com.datastax.internal.com_google_common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:41)
    at com.datastax.internal.com_google_common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:77)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
    at java.base/java.lang.Thread.run(Thread.java:832)
[cqltabular2_default_schema:001] WARN com.datastax.cql3.scylla_shaded.driver.core.Session - Error creating pool to /10.0.2.65:9042
com.datastax.cql3.scylla_shaded.driver.core.exceptions.OperationTimedOutException: [/10.0.2.65:9042] Operation timed out
    at com.datastax.cql3.scylla_shaded.driver.core.Connection$Future.onTimeout(Connection.java:1642)
    at com.datastax.cql3.scylla_shaded.driver.core.Connection$ResponseHandler$1.run(Connection.java:1721)
    at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:672)
    at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:747)
    at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:472)
    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
    at java.base/java.lang.Thread.run(Thread.java:832)

```

Another occurrence on `scylla-driver-core 3.11.5.11`, seen in Scylla Jenkins on March 29, 2026:

Build: https://jenkins.scylladb.com/job/scylla-master/job/dtest-debug/558/
Failing test: https://jenkins.scylladb.com/job/scylla-master/job/dtest-debug/558/testReport/junit/replace_address_test/TestReplaceAddress/Run_Dtest_Parallel_Cloud_Machines___Dtest___full_split005___test_replace_with_background_workload_rbo_enabled_/

Relevant stack:
```text
ERROR [cluster1-worker-0] 2026-03-29 08:54:49,544 ExceptionCatchingRunnable.java:36 - Unexpected error while executing task
java.lang.NullPointerException: Cannot read the array length because "<local2>" is null
    at com.datastax.driver.core.HostConnectionPool.discardAvailableConnections(HostConnectionPool.java:1016)
    at com.datastax.driver.core.HostConnectionPool.closeAsync(HostConnectionPool.java:987)
    at com.datastax.driver.core.SessionManager.removePool(SessionManager.java:446)
    at com.datastax.driver.core.SessionManager.onDown(SessionManager.java:534)
    at com.datastax.driver.core.Cluster$Manager.onDown(Cluster.java:2165)
```

The surrounding log shows topology churn during node replacement:
- `/127.0.71.3:9042` is discovered and then goes down
- pool creation to `/127.0.71.3:9042` fails with `Connection refused`
- host-down cleanup then goes through `removePool -> closeAsync -> discardAvailableConnections`

The same run also shows:
```text
WARN  [main] ... Error creating pool to /127.0.71.3:9042
com.datastax.driver.core.exceptions.TransportException: [/127.0.71.3:9042] Cannot connect
Caused by: java.net.ConnectException: Connection refused
```

This matches the analysis already described here: if the initial pool connection fails before `initAsyncWithConnection()` initializes `connections` / `trash`, then `closeAsync()` can NPE on cleanup.

In this Jenkins run the test ultimately failed for a separate `cassandra-stress` schema/rack issue, so this looks like a secondary driver crash rather than the primary failure, but it is still reproducible noise from real topology changes.

--- a/driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java
+++ b/driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java
@@ -301,18 +301,24 @@ class HostConnectionPool implements Connection.Owner {
         maxConnections / shardsCount + (maxConnections % shardsCount > 0 ? 1 : 0);
     int toCreate = shardsCount * connectionsPerShard;
 
-    this.connections = new List[shardsCount];
-    scheduledForCreation = new AtomicInteger[shardsCount];
-    open = new AtomicInteger[shardsCount];
-    trash = new Set[shardsCount];
-    pendingBorrows = new Queue[shardsCount];
+    List<Connection>[] connectionsByShard = new List[shardsCount];
+    AtomicInteger[] scheduledForCreationByShard = new AtomicInteger[shardsCount];
+    AtomicInteger[] openByShard = new AtomicInteger[shardsCount];
+    Set<Connection>[] trashByShard = new Set[shardsCount];
+    Queue<PendingBorrow>[] pendingBorrowsByShard = new Queue[shardsCount];
     for (int i = 0; i < shardsCount; ++i) {
-      this.connections[i] = new CopyOnWriteArrayList<Connection>();
-      scheduledForCreation[i] = new AtomicInteger();
-      open[i] = new AtomicInteger();
-      trash[i] = new CopyOnWriteArraySet<Connection>();
-      pendingBorrows[i] = new ConcurrentLinkedQueue<PendingBorrow>();
+      connectionsByShard[i] = new CopyOnWriteArrayList<Connection>();
+      scheduledForCreationByShard[i] = new AtomicInteger();
+      openByShard[i] = new AtomicInteger();
+      trashByShard[i] = new CopyOnWriteArraySet<Connection>();
+      pendingBorrowsByShard[i] = new ConcurrentLinkedQueue<PendingBorrow>();
     }
+
+    this.connections = connectionsByShard;
+    scheduledForCreation = scheduledForCreationByShard;
+    open = openByShard;
+    trash = trashByShard;
+    pendingBorrows = pendingBorrowsByShard;
 
     final List<Connection> connections = Lists.newArrayListWithCapacity(toCreate);
     final List<ListenableFuture<Void>> connectionFutures =
@@ -995,44 +1001,75 @@ class HostConnectionPool implements Connection.Owner {
   }
 
   int opened() {
+    AtomicInteger[] open = this.open;
+    if (open == null) {
+      return 0;
+    }
+
     int result = 0;
     for (AtomicInteger o : open) {
-      result += o.get();
+      if (o != null) {
+        result += o.get();
+      }
     }
     return result;
   }
 
   int trashed() {
+    Set<Connection>[] trash = this.trash;
+    if (trash == null) {
+      return 0;
+    }
+
     int size = 0;
     for (final Set<Connection> shardConnections : trash) {
-      size += shardConnections.size();
+      if (shardConnections != null) {
+        size += shardConnections.size();
+      }
     }
     return size;
   }
 
   private List<CloseFuture> discardAvailableConnections() {
-    // Note: if this gets called before initialization has completed, both connections and trash
-    // will be empty,
-    // so this will return an empty list
+    List<Connection>[] connections = this.connections;
+    Set<Connection>[] trash = this.trash;
+    if (connections == null || trash == null) {
+      return new ArrayList<CloseFuture>(0);
+    }
+
+    // Note: if this gets called before initialization has completed, connections and trash might
+    // still be partially populated, so null entries are ignored and this still returns a valid list
 
     int size = 0;
     for (final Set<Connection> shardConnections : trash) {
-      size += shardConnections.size();
+      if (shardConnections != null) {
+        size += shardConnections.size();
+      }
     }
     for (final List<Connection> shardConnections : connections) {
-      size += shardConnections.size();
+      if (shardConnections != null) {
+        size += shardConnections.size();
+      }
     }
     List<CloseFuture> futures = new ArrayList<CloseFuture>(size);
 
     for (final List<Connection> shardConnections : connections) {
+      if (shardConnections == null) {
+        continue;
+      }
       for (final Connection connection : shardConnections) {
         CloseFuture future = connection.closeAsync();
         future.addListener(
             new Runnable() {
               @Override
               public void run() {
-                if (connection.state.compareAndSet(OPEN, GONE)) {
-                  open[connection.shardId()].decrementAndGet();
+                AtomicInteger[] open = HostConnectionPool.this.open;
+                AtomicInteger openCount =
+                    open != null && connection.shardId() < open.length
+                        ? open[connection.shardId()]
+                        : null;
+                if (openCount != null && connection.state.compareAndSet(OPEN, GONE)) {
+                  openCount.decrementAndGet();
                 }
               }
             },
@@ -1043,6 +1080,9 @@ class HostConnectionPool implements Connection.Owner {
 
     // Some connections in the trash might still be open if they hadn't reached their idle timeout
     for (final Set<Connection> shardConnections : trash) {
+      if (shardConnections == null) {
+        continue;
+      }
       for (final Connection connection : shardConnections) {
         futures.add(connection.closeAsync());
       }

--- a/driver-core/src/test/java/com/datastax/driver/core/HostConnectionPoolTest.java
+++ b/driver-core/src/test/java/com/datastax/driver/core/HostConnectionPoolTest.java
@@ -300,6 +300,26 @@ public class HostConnectionPoolTest extends ScassandraTestBase.PerClassCluster {
     }
   }
 
+  @Test(groups = "short")
+  public void should_close_uninitialized_pool_without_npe() throws Exception {
+    Cluster cluster = createClusterBuilder().build();
+    try {
+      Session session = cluster.connect();
+      Host host = TestUtils.findHost(cluster, 1);
+      SessionManager sessionManager = (SessionManager) session;
+      HostConnectionPool pool = new HostConnectionPool(host, HostDistance.LOCAL, sessionManager);
+
+      CloseFuture closeFuture = pool.closeAsync();
+
+      closeFuture.get(5, TimeUnit.SECONDS);
+      assertThat(pool.isClosed()).isTrue();
+      assertThat(pool.opened()).isEqualTo(0);
+      assertThat(pool.trashed()).isEqualTo(0);
+    } finally {
+      cluster.close();
+    }
+  }
+
   /**
    * Validates that if the keyspace tied to the Session's pool state is different than the keyspace
    * on the connection being used in dequeue that {@link Connection#setKeyspaceAsync(String)} is set

